drkafka/src/main/java/com/pinterest/doctorkafka/KafkaClusterManager.java

```diff
@@ -594,17 +593,37 @@ public UnderReplicatedReason getUnderReplicatedReason(String brokerHost,
 
   /**
    * Generate reassignment plan for dead brokers
+   * Different failure handling for locality-aware and non-locality-aware:
+   * - Locality-aware: will reassign topicPartitions that succeeded to get assigned to another
+   *   same-locality broker, but will send alert notifying that partitions that failed (due to
+   *   insufficient capacity on the same locality).
+   * - Non-locality-aware: current reassignment will fail (reassignments are all-or-none).
    */
   private Map<TopicPartition, Integer[]> generateReassignmentPlanForDeadBrokers(
       List<OutOfSyncReplica> outOfSyncReplicas) {
     Map<TopicPartition, Integer[]> replicasMap = new HashMap<>();
     boolean success = true;
+    boolean isLocalityAware = clusterConfig.enabledRackAwareness();
+
+    Map<String, PriorityQueue<KafkaBroker>> brokerQueueByLocality = null;
+    PriorityQueue<KafkaBroker> brokerQueue = null;
+    Map<OutOfSyncReplica, Map<Integer, String>> reassignmentToLocalityFailures = new HashMap<>();
+    if(isLocalityAware){
+      brokerQueueByLocality = kafkaCluster.getBrokerQueueByLocality();
+    } else {
+      brokerQueue = kafkaCluster.getBrokerQueue();
+    }
 
-    PriorityQueue<KafkaBroker> brokerQueue = kafkaCluster.getBrokerQueue();
     for (OutOfSyncReplica oosReplica : outOfSyncReplicas) {
-      Map<Integer, KafkaBroker> replacedNodes =
-          kafkaCluster.getAlternativeBrokers(brokerQueue, oosReplica);
-      if (replacedNodes == null) {
+      Map<Integer, KafkaBroker> replacedNodes;
+      replacedNodes = isLocalityAware
+                      ? kafkaCluster.getAlternativeBrokersByLocality(
+                          brokerQueueByLocality,
+                          oosReplica,
+                          reassignmentToLocalityFailures
+                        )
+                      : kafkaCluster.getAlternativeBrokers(brokerQueue, oosReplica);
+      if (replacedNodes == null && !isLocalityAware) {
```

Comment:
@yuyang08 The difference between locality-aware and non-locality-aware is introduced here:
If locality-awareness is disabled, success will get marked as false and the entire reassignment will get abandoned.
If locality-awareness is enabled, the reassignment for this oosReplica will fail but we will still try to reassign other oosReplicas.

Reply:
will we get null pointer exception from line 638 when `isLocalityAware` is set to true, and `replaceNodes == null`? 

    newReplicas[i] = replacedNodes.containsKey(brokerId) ? replacedNodes.get(brokerId).id()
                                                               : brokerId;

Reply:
@yuyang08 yes, fixed in latest commit